### PR TITLE
Readme for Z algorithm

### LINKED ISSUE #2478: Z algorithm
- [x] C
- [x] C++
- [ ] C# - @HardevKhandhar
- [x] CoffeeScript
- [ ] Dart - @vanya-24
- [ ] Go - @KamalDGRT
- [x] Java
- [x] JS - @abhu-A-J
- [ ] Kotlin
- [ ] Php - @shivanidalmia1203
- [x] Python
- [x] Readme - @syedareehaquasar
- [x] Ruby
- [x] Rust - @jedi-starlord
- [ ] TypeScript

### PR DESCRIPTION
**Pull request by syedareehaquasar**
Fixes #2478 

#### Changes proposed in this pull request:
- readme for #2478  Z algorithm

#### Files Added:

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by syedareehaquasar** [comment]
plese review @jainaman224

File: Ford_Fulkerson_Method/README.md [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+∗
+  is the maximum flow of the network. The Ford-Fulkerson algorithm was eventually improved upon by the Edmonds-Karp algorithm, which does the same thing in O\big(V^2 \cdot E\big)O(V 
+2
+ ⋅E) time, independent of the maximum flow value.
+
+# Time Complexity
+O(max_flow * E)
+Time complexity of the above algorithm is O(max_flow * E). We run a loop while there is an augmenting path. In worst case, we may add 1 unit flow in every iteration. Therefore the time complexity becomes O(max_flow * E)
+
+# Implementation
```
Comment (asha15) [inline Ford_Fulkerson_Method/README.md]: The two heading needs to come at the end of the page

File: Ford_Fulkerson_Method/README.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+ - # [coffee code]
+ - # [cpp code]
+ - # [c# code]
+ - # [go code]
+ - # [java code]
+ - # [javaScript code]
+ - # [kotlin code]
+ - # [python code]
+ - # [ruby code]
+ - # [dart code]
```
Comment (asha15) [inline Ford_Fulkerson_Method/README.md]: Remove #

File: Ford_Fulkerson_Method/README.md [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+
```
Comment (asha15) [inline Ford_Fulkerson_Method/README.md]: remove these lines

File: Ford_Fulkerson_Method/README.md [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        else:
+            flow(u, v) = flow(u, v) - residual_capacity(p)
+return flow
+```
+
+
+
+python implementation:
+
+```
```
Comment (asha15) [inline Ford_Fulkerson_Method/README.md]: no need of the implementation

File: Ford_Fulkerson_Method/README.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+initialize flow to 0
+path = findAugmentingPath(G, s, t)
+while path exists:
+    augment flow along path                 #This is purposefully ambiguous for now
+    G_f = createResidualGraph()
+    path = findAugmentingPath(G_f, s, t)
+return flow
+```
+
+## ALGORITHM
```
Comment (asha15) [inline Ford_Fulkerson_Method/README.md]: It's better if you provide steps for the algorithm

**Comment by syedareehaquasar** [comment]
@asha15  please review updated readme ... all necessary changes done

**Comment by syedareehaquasar** [comment]
@asha15 please review updated readme ... all necessary changes done

File: Z_Algorithm/README.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+Z array for above concatenated string is {x, 1, 0, 0, 0, 
+                                          3, 1, 0, 2, 1}.
+Since length of pattern is 3, the value 3 in Z array 
+indicates presence of pattern. 
+```
+
+### How to construct Z array?
+A Simple Solution is two run two nested loops, the outer loop goes to every index and the inner loop finds length of the longest prefix that matches substring starting at current index. 
+
+# ALGORITHM
```
Comment (asha15) [inline Z_Algorithm/README.md:41]: This a huge description . remove it and provide some simple steps in bullet form

File: Z_Algorithm/README.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+ - # [coffee code](https://github.com/jainaman224/Algo_Ds_Notes/blob/master/Z_Algorithm/Z_Algorithm.coffee)
+ - # [cpp code](https://github.com/jainaman224/Algo_Ds_Notes/blob/master/Z_Algorithm/Z_Algorithm.cpp)
+ - # [c# code]
+ - # [go code]
+ - # [java code](https://github.com/jainaman224/Algo_Ds_Notes/blob/master/Z_Algorithm/Z_Algorithm.java)
+ - # [javaScript code](https://github.com/jainaman224/Algo_Ds_Notes/blob/master/Z_Algorithm/Z_Algorithm.js)
+ - # [kotlin code]
+ - # [python code](https://github.com/jainaman224/Algo_Ds_Notes/blob/master/Z_Algorithm/Z_Algorithm.py)
+ - # [ruby code](https://github.com/jainaman224/Algo_Ds_Notes/blob/master/Z_Algorithm/Z_Algorithm.rb)
+ - # [dart code]
```
Comment (asha15) [inline Z_Algorithm/README.md:71]: Don't make this as a heading.
remove the #

File: Ford_Fulkerson_Method/README.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
+# Ford-Fulkerson Algorithm
+
```
Comment (asha15) [inline Ford_Fulkerson_Method/README.md:2]: remove this file

File: Z_Algorithm/README.md (hunk 6 above)
Comment (asha15) [inline Z_Algorithm/README.md:41]: make the heading to ##

File: Z_Algorithm/README.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+Otherwise, i ≤ R, so the current [L, R] extends at least to i. Let k = i - L. We know that Z[i] ≥ min(Z[k], R - i + 1) because S[i...] matches S[k...] for at least R - i + 1 characters (they are in the [L, R] interval which we know to be a prefix-substring). Now we have a few more cases to consider.
+If Z[k] < R - i + 1, then there is no longer prefix-substring starting at S[i] (or else Z[k] would be larger), meaning Z[i] = Z[k] and [L, R] stays the same. The latter is true because [L, R] only changes if there is a prefix-substring starting at S[i] that extends beyond R, which we know is not the case here.
+If Z[k] ≥ R - i + 1, then it is possible for S[i...] to match S[0...] for more than R - i + 1 characters (i.e. past position R). Thus we need to update [L, R] by setting L = i and matching from S[R + 1] forward to obtain the new R. Again, we get Z[i] during this.
+The process computes all of the Z values in a single pass over the string, so we're done. Correctness is inherent in the algorithm and is pretty intuitively clear.
+
+
+# Time Complexity
```
Comment (asha15) [inline Z_Algorithm/README.md:55]: making this as a h2

File: Z_Algorithm/README.md [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+O(n)
+The algorithm runs in O(n) time. Characters are never compared at positions less than , and every time a match is found,  is increased by one, so there are at most n comparisons
+
+# Implementation
```
Comment (asha15) [inline Z_Algorithm/README.md:59]: make this as a h2

**Comment by syedareehaquasar** [comment]
@asha15 please review

File: Z_Algorithm/README.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
+# Z algorithm
```
Comment (asha15) [inline Z_Algorithm/README.md:1]: define the heading in upper case letters

File: Z_Algorithm/README.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    z[i] = R-L; R--;
+  } else {
+    int k = i-L;
+    if (z[k] < R-i+1) z[i] = z[k];
+    else {
+      L = i;
+      while (R < n && s[R-L] == s[R]) R++;
+      z[i] = R-L; R--;
+    }
+  }
```
Comment (asha15) [inline Z_Algorithm/README.md:62]: correct the indentation

File: Z_Algorithm/README.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+}
+```
+
```
Comment (asha15) [inline Z_Algorithm/README.md:65]: remove these lines

**Comment by syedareehaquasar** [comment]
@asha15  please check done squash too

**Comment by syedareehaquasar** [comment]
please review @asha15

**Pull request merged**